File: backend/app/distribution/publisher.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any

from app.distribution.redis_bus import publish


class CanalDistribucion(ABC):
    @abstractmethod
    def enviar(self, topico: str, mensaje: dict[str, Any]) -> None: ...


class RedisChannel(CanalDistribucion):
    def enviar(self, topico: str, mensaje: dict[str, Any]) -> None:
        publish(topico, mensaje)


class APIRest(CanalDistribucion):
    def enviar(self, topico: str, mensaje: dict[str, Any]) -> None: ...


class Publisher:
    def __init__(self) -> None:
        self._canales: list[CanalDistribucion] = []

    def suscribir(self, canal: CanalDistribucion) -> None:
        self._canales.append(canal)

    def publicar(self, topico: str, mensaje: dict[str, Any]) -> None:
        for canal in self._canales:
            canal.enviar(topico, mensaje)


publisher = Publisher()
publisher.suscribir(RedisChannel())
publisher.suscribir(APIRest())

File: backend/src/main/java/dev/solomillo/distribution/Publisher.java
package dev.solomillo.distribution;

import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class Publisher {

    private final List<CanalDistribucion> canales;

    public Publisher(List<CanalDistribucion> canales) {
        this.canales = canales;
    }

    public void publicar(String topico, Object mensaje) {
        canales.forEach(c -> c.enviar(topico, mensaje));
    }
}
